### DISCUSSION TITLE
Cannot load api key from ~\.authinfo

### DISCUSSION THREAD

**Discussion by tsoernes** [karthink/gptel#1380] [Q&A]
I have this config in doom emacs
```elisp
(setq-default gptel-model 'm-gpt-4o-mini-18072024)
(setq gptel-backend (gptel-make-azure "Azure-1-GPT-4o-mini"
                      :protocol "https"
                      :host "<MY_HOST>.openai.azure.com"
                      :endpoint "/openai/deployments/m-gpt-4o-mini-18072024/chat/completions?api-version=2024-08-01-preview"
                      :stream t
                      :key #'gptel-api-key-from-auth-source
                      :models '(m-gpt-4o-mini-18072024)))
```
Is this correct?
Cause I have the file
`~/.authinfo` with
```
machine api.openai.com login apikey password <MY_TOKEN>
```
But I'm asked for a GPT API key every time I open `gptel`.

**Comment by karthink** [comment]
What happens if you run `(gptel-api-key-from-auth-source)`?  Does it return the key?

**Comment by tsoernes** [comment]
if: No ‘gptel-api-key’ found in the auth source

**Comment by tschwinge** [comment]
I suppose the problem is that your `~/.authinfo`'s `machine api.openai.com` doesn't match your `gptel-backend`'s `:host "<MY_HOST>.openai.azure.com"`?

**Comment by karthink** [comment]
@tsoernes `gptel-api-key-from-auth-source` looks up the `:host` in your authinfo.  So the authinfo entry should look like

```
machine <MY_HOST>.openai.azure.com login apikey password <MY_TOKEN>
````

OR

You can write your own function that looks up the right key in your authinfo and use that in place of `gptel-api-key-from-auth-source`.

**Comment by glubbfreund** [comment]
im not sure If I should answer here but I have the same Issue except with ChatGPT. Do you use windows? It doesnt find the file in windows but within WSL2 it works with just the same config and .authinfo file. Seems to be a path issue on my side. 

**It works on both systems if I use the following approach:**
```
(require 'gptel)
(setq auth-sources '("~/.authinfo"))
(setq gptel-api-key (auth-source-pick-first-password :host "api.openai.com"))
```

**Comment by karthink** [comment]
@tsoernes Does my above reply work for you?

**Comment by vikdotdev** [comment]
I had the same problem on doom emacs and was not able to fully resolve it. On initial startup it would never read or prompt for the gpg password, but when I cleared the authinfo cache with `(auth-source-forget-all-cached)` and ran the authinfo read function again, it successfully read and stored the key.

On my custom configuration it worked without any issues, which makes me think there's something that doom does in how it works with desktop environment password prompts.

**Comment by pixelsandpointers** [comment]
> im not sure If I should answer here but I have the same Issue except with ChatGPT. Do you use windows? It doesnt find the file in windows but within WSL2 it works with just the same config and .authinfo file. Seems to be a path issue on my side.
> 
> **It works on both systems if I use the following approach:**
> 
> ```
> (require 'gptel)
> (setq auth-sources '("~/.authinfo"))
> (setq gptel-api-key (auth-source-pick-first-password :host "api.openai.com"))
> ```

I had the same problem with the config for Doom. The above code solved it.

**Comment by andy-novikov** [comment]
I have this in doom `config.el`, that works.
```lisp
(use-package! gptel
  :config
  (setq gptel-model   'deepseek-chat
        (gptel-make-openai "DeepSeek"     ;Any name you want
          :host "api.deepseek.com"
          :endpoint "/chat/completions"
          :key (auth-source-pick-first-password :host "api.deepseek.com")
          :models '(deepseek-chat deepseek-reasoner)))
  )
```

I think I had to enable
```lisp
       pass              ; password manager for nerds
```
in `init.el` for this to work.

**Comment by rvadhavk** [comment]
I'm not sure if this is what others are running into, but I made a mistake when first populating my `~/.authinfo`, and Emacs wouldn't pick up the corrections until I ran `auth-source-forget-all-cached`.

**Comment by Keisn1** [comment]
I don't know why, but auth-source-search just couldn't get the key when "apikey" was indicated as user (as it is inside gptel-api-key-from-source). Probably it is because I have other keys with "apikey" as user inside my .authinfo. 
So I have the following solution

```
(gptel-make-anthropic "Claude"       
  :stream t                            
  :key (lambda () (gptel-api-key-from-auth-source "api.anthropic.com" "something_else_than_apikey")))
```

**Comment by cartoonist** [comment]
I have the same problem. I couldn't find out exactly why `auth-source` returns `nil` without clearing the cache by `(auth-source-forget-all-cached)`.

**UPDATE:** I just found out that `(auth-source-search :host ...)` treats my `.authinfo.gpg` file as unencrypted and returns gibberish as a plist. That's why calling `auth-info-password` on the output return `nil`. 

This is my workaround that solved the issue for me:

```lisp
(use-package! gptel
  :after auth-source
  :config
  (setq! gptel-api-key (lambda ()
                         (progn
                           (auth-source-forget-all-cached)
                           (auth-source-pick-first-password :host "api.openai.com"))))
  (gptel-make-anthropic "Claude"
    :stream t             ; Streaming responses
    :key (lambda ()
           (progn
             (auth-source-forget-all-cached)
             (auth-source-pick-first-password :host "console.anthropic.com")))))
```

**Comment by iftheshoefritz** [comment]
None of the workarounds above are working for me. Maybe I'm being confused by translating to spacemacs?

In `.spacemacs`:

```
  dotspacemacs-configuration-layers
  '(
       (llm-client :variables
                 llm-client-enable-gptel t)
  )
  ...
  ((defun dotspacemacs/user-config ()
  (with-eval-after-load 'gptel
    (message "testing auth-source after load: %s"
             (auth-source-pick-first-password :host "api.anthropic.com"))
    (message "testing gptel key after auth-source load: %s" (gptel-api-key-from-auth-source "api.anthropic.com" "apikey"))

    ;; slightly different from the docs but the docs as stated don't work either
    (gptel-make-anthropic "Claude" :stream t :key (gptel-api-key-from-auth-source "api.anthropic.com" "apikey"))

    )
```

results in this output when I load my config:

```
testing auth-source after load: sk-ant-xxxxx
testing gptel key after auth-source load: sk-ant-xxxxx
```

Unfortunately `gptel-send` still gives me: `No ‘gptel-api-key’ found in the auth source`
